### PR TITLE
Python 3.12 Support

### LINKED ISSUE #818: Run CI tests against Python 3.12.0 release candidate
### What should we change and why?

[Python 3.12.0rc1](https://www.python.org/downloads/release/python-3120rc1/) has been released. We'll need a supporting [numpy release](https://github.com/numpy/numpy/issues/23808#issuecomment-1670277265) and then we should run our CI tests against 3.12.

### PR DESCRIPTION
**Pull request by austinmatherne-wk**
#### Reason for change
Resolves #818

#### Description of change
Update Arelle to be compatible with Python 3.12 and workflows (besides frozen builds) to use it.
cx_Freeze doesn't support 3.12 yet, so frozen builds will continue to use 3.11 for the time being, but cx_Freeze was updated to the latest release to support newer versions of Matplotlib and NumPy that support Python 3.12.

Updating cx_Freeze broke the code signing step of our macOS build workflow as the library recently overhauled their macOS builds. Fortunately a large part of the overhaul was to make code signing work from within the library, so I was able to delete our code signing step and configure cx_Freeze to handle signing instead.

Python versions prior to 3.12 incorrectly mapped the `C` locale to `en_US` this necessitated overhauling local handling in Arelle. Previously locales and preferred label languages were intermingled (`lang.replace('-', '_')` and `lang.replace('_', '-')` used all over the place) when there's no reason that preferred label languages should be limited to system locales.

#### Steps to Test
* CI
* Frozen builds already tested for [Linux](https://github.com/austinmatherne-wk/Arelle/actions/runs/7394873206), [macOS](https://github.com/austinmatherne-wk/Arelle/actions/runs/7394873849), and [Windows](https://github.com/austinmatherne-wk/Arelle/actions/runs/7394874508).

**review**:
@Arelle/arelle

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

**Review by derekgengenbacher-wf — DISMISSED** [review]

File: arelle/Locale.py [hunk 1]
```diff
@@ -98,23 +86,145 @@ def getUserLocale(localeCode: str = '') -> tuple[LocaleDict, str | None]:
...
+    Returns a list of language code formats that can be used with gettext to look up translation files.
+    [en_US, en-US, en]
+    [fr]
+    """
     if lang is None:
         lang = getLanguageCode()
-    # allow searching on the lang with country part, either python or standard form, or just language
-    return [lang, lang.replace("-","_"), lang.partition("-")[0]]
+    if LANGUAGE_CODE_LOCALE_SEPARATOR in lang:
+        return [lang, langCodeXBRLFormat(lang), lang.partition(LANGUAGE_CODE_LOCALE_SEPARATOR)[0]]
```
Comment (brettkail-wk) [inline arelle/Locale.py:106]: I'm a bit confused about the terminology here.  I think the relevant string formats are "POSIX locale" and BCP 47 "language tag".  Is the latter what you're calling `LANGUAGE_CODE_XML_SEPARATOR` / `langCodeXBRLFormat`?  If so, perhaps `LANGUAGE_TAG_SEPARATOR`, `langTagFormat`, etc. would be clearer names.

The `lang` param sounds like it's language tag, but the code checks for a locale string separator, and the caller is passing a var named `locale`, which seems to come from a user-specified string.  Maybe it's really just a fuzzy config string that might be lang or locale and `configLang` would be clearer, or perhaps just a small doc comment that the string is user-specified, so we're making a best effort.

Comment (austinmatherne-wk) [reply arelle/Locale.py]: I updated a bunch of the variable names and included comments. It should be much more clear now.

**Review by derekgengenbacher-wf — APPROVED** [review]

**Review by aaroncameron-wk — APPROVED** [review]

**Comment by aaroncameron-wk** [comment]
QA +10
CI passes.

**Pull request merged**